include/dr/mp/algorithms/count.hpp

```diff
@@ -0,0 +1,159 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+namespace dr::mp::__detail {
+
+inline auto add_counts(rng::forward_range auto &&r) {
+  rng::range_difference_t<decltype(r)> zero{};
+
+  return std::accumulate(rng::begin(r), rng::end(r), zero);
+}
+
+inline auto std_count_if(rng::forward_range auto &&r, auto &&pred) {
+  using count_type = rng::range_difference_t<decltype(r)>;
+
+  if (rng::empty(r)) {
+    return count_type{};
+  }
+
+  return std::count_if(std::execution::par_unseq,
+                       dr::__detail::direct_iterator(rng::begin(r)),
+                       dr::__detail::direct_iterator(rng::end(r)), pred);
+}
+
+inline auto dpl_count_if(rng::forward_range auto &&r, auto &&pred) {
+  using count_type = rng::range_difference_t<decltype(r)>;
+
+#ifdef SYCL_LANGUAGE_VERSION
+  if (rng::empty(r)) {
+    return count_type{};
+  }
+
+  return std::count_if(mp::dpl_policy(),
+                       dr::__detail::direct_iterator(rng::begin(r)),
+                       dr::__detail::direct_iterator(rng::end(r)), pred);
+#else
+  assert(false);
+  return count_type{};
+#endif
+}
+
+template <dr::distributed_range DR>
+auto count_if(std::size_t root, bool root_provided, DR &&dr, auto &&pred) {
+  using count_type = rng::range_difference_t<decltype(dr)>;
+  auto comm = mp::default_comm();
+
+  if (rng::empty(dr)) {
+    return count_type{};
+  }
+
+  if (aligned(dr)) {
+    dr::drlog.debug("Parallel count\n");
+
+    // Count within the local segments
+    auto count = [=](auto &&r) {
```

Comment:
Consider moving the lambda outside the count_if function

Reply:
Some kind of lambda still has to stay here as we need something to pass to `rng::views::transform`. If all of the lambda's code were moved to a separate function, then it would have to take a predicate as a parameter, which would in turn require a new lambda with its only purpose being to pass `pred` to that function. I don't really see a nice solution that would nicely move the lambda outside, unless I'm missing something here